### DISCUSSION TITLE
Confused about the `z.enum` argument type

### DISCUSSION THREAD

**Discussion by kasbah** [colinhacks/zod#1063] [Q&A] [answered]
I'm a bit confused about the `z.enum` argument type. Let's say I have

tsconfig.json
```json
{
  "compilerOptions": {
    "lib": ["ES2021"],
    "module": "commonjs",
    "target": "ES2021",
    "strict": true,
    "noEmit": true,
    "esModuleInterop": true,
    "resolveJsonModule": true
  },
  "files": ["example.ts"]
}
```
keys.json
```json
["key1", "key2"]
```
example.ts
```ts
import { z } from "zod";
import keys from "./keys.json";

const Keys = z.enum(keys);

console.log(Keys.options);
```

I get this type error when running `tsc` (Version 4.2.4)

```
example.ts:4:21 - error TS2769: No overload matches this call.
  Overload 1 of 2, '(values: readonly [string, ...string[]]): ZodEnum<[string, ...string[]]>', gave the following error.
    Argument of type 'string[]' is not assignable to parameter of type 'readonly [string, ...string[]]'.
      Source provides no match for required element at position 0 in target.
  Overload 2 of 2, '(values: [string, ...string[]]): ZodEnum<[string, ...string[]]>', gave the following error.
    Argument of type 'string[]' is not assignable to parameter of type '[string, ...string[]]'.
      Source provides no match for required element at position 0 in target.

4 const Keys = z.enum(keys);
                      ~~~~

```

Which is sort of expected from what it says in the zod readme. But following the cues of some of the discussions/issues here I can do:

```ts
const Keys = z.enum([keys[0], ...keys.slice(1)]);
```

And that type checks fine and seems to work. Can someone explain to me why the type is like this and what the difference is between the two? Could relying on this cause issues down the line? 

I guess the confusing thing to me is, why doesn't this type check? 

```ts
const Keys = z.enum([...keys]);
```

**Comment by MatteoGauthier** [comment]
```ts
const Keys = z.enum(keys as unknown as readonly [string, ...string[]]);
```

**Reply by kasbah** [reply]
Thanks for the answer. That type is a bit less confusing for sure. I'll use that.

Do you have any insight into why the type is like this. Is this just how you type a non-empty Array in Typescript?

**Reply by MatteoGauthier** [reply]
To be honest that's just a snippet I wrote rapidly, I have not been further

**Comment by ibrandao-godaddy** [comment]
I wouldn't use `as unknown` cast because it forfeits type safety. The proper answer for this problem is on this issue: https://github.com/colinhacks/zod/issues/73

**Reply by kasbah** [reply]
Thanks, that issue is quite long though. What's the answer?

**Reply by ibrandao-godaddy** [reply]
Use `as const` or `as [string, ...string[]]`

**Comment by JacobWeisenburger** [comment, marked as answer]
Zod Enum requires more than 1 item in the string tuple. Not sure why it was designed this way, but it appears to be intentional:

https://github.com/colinhacks/zod/blob/38fac35a63b143666412d1f3f267bbecb04908b5/src/types.ts#L3982-L4005

**Reply by kasbah** [reply]
Well, it does make sense: what's an enum of zero elements exactly? I think at the time I was just confused by the `[string, ...string[]]` type because it is a bit of an un-intuitive way to make sure there's one element, but I guess that's how you do it in TS.
